File: frontend-app/src/features/candidate/candidateService.js
import axiosClient from '../../services/axiosClient';

const candidateService = {
  // ==========================================
  // 1. QUẢN LÝ CV (giữ nguyên như cũ)
  // ==========================================

  createCv: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/cvs/generated`, payload);
    return response.data;
  },

  uploadCv: async (userId, file) => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await axiosClient.post(`/candidates/${userId}/cvs`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  getCvs: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/cvs`);
    return response.data;
  },

  getMyCVs: async (candidateId) => {
    const response = await axiosClient.get(`/candidates/${candidateId}/cvs`);
    return response.data;
  },

  getCvDetail: async (userId, cvId) => {
    const response = await axiosClient.get(`/candidates/${userId}/cvs/${cvId}`);
    return response.data;
  },

  updateCv: async (userId, cvId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/cvs/generated/${cvId}`, payload);
    return response.data;
  },

  renameCv: async (userId, cvId, newName) => {
    const response = await axiosClient.put(`/candidates/${userId}/cvs/${cvId}/rename`, null, {
      params: { newName }
    });
    return response.data;
  },

  uploadCvThumbnail: async (userId, cvId, file) => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await axiosClient.post(`/candidates/${userId}/cvs/${cvId}/thumbnail`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  deleteCv: async (userId, cvId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/cvs/${cvId}`);
    return response.data;
  },

  // ==========================================
  // 2. QUẢN LÝ PROFILE (giữ nguyên như cũ + avatar mới)
  // ==========================================

  getProfile: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/profile`);
    return response.data;
  },

  createOrUpdateProfile: async (userId, profileData) => {
    const response = await axiosClient.post(`/candidates/${userId}/profile`, profileData);
    return response.data;
  },

  uploadAvatar: async (userId, file) => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await axiosClient.post(`/candidates/${userId}/avatar`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  // [MỚI] Lấy toàn bộ ProfilePage (layout + 9 block danh sách) trong 1 lần gọi — Tiến độ 7
  getFullProfile: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/profile/full`);
    return response.data;
  },

  // ==========================================
  // 3. PROFILE LAYOUT (SANDBOX KÉO-THẢ) — [MỚI] Tiến độ 4
  // ==========================================

  getProfileLayout: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/profile/layout`);
    return response.data;
  },

  // items: [{ blockType, position }, ...]
  reorderProfileLayout: async (userId, items) => {
    const response = await axiosClient.put(`/candidates/${userId}/profile/layout/reorder`, { items });
    return response.data;
  },

  toggleBlockVisibility: async (userId, blockType, visible) => {
    const response = await axiosClient.patch(
      `/candidates/${userId}/profile/layout/${blockType}/visibility`,
      { visible }
    );
    return response.data;
  },

  // ==========================================
  // 4. QUẢN LÝ SKILLS (giữ nguyên + bổ sung reorder/assign/remove)
  // ==========================================

  getSkills: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/skills`);
    return response.data;
  },

  createSkill: async (userId, skillData) => {
    const response = await axiosClient.post(`/candidates/${userId}/skills`, skillData);
    return response.data;
  },

  updateSkill: async (userId, skillId, skillData) => {
    const response = await axiosClient.put(`/candidates/${userId}/skills/${skillId}`, skillData);
    return response.data;
  },

  // [MỚI] Gán nhanh 1 skill đã tồn tại (chọn từ dropdown), không cần nhập tên
  assignSkill: async (userId, skillId) => {
    const response = await axiosClient.post(`/candidates/${userId}/skills/${skillId}/assign`);
    return response.data;
  },

  removeSkill: async (userId, skillId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/skills/${skillId}`);
    return response.data;
  },

  // [MỚI] Kéo-thả sắp xếp lại thứ tự skill — Tiến độ 6
  // items: [{ skillId, displayOrder }, ...]
  reorderSkills: async (userId, items) => {
    const response = await axiosClient.put(`/candidates/${userId}/skills/reorder`, { items });
    return response.data;
  },

  // ==========================================
  // 5. DROPDOWN SKILL/LANGUAGE + ĐỀ XUẤT ADMIN — [MỚI] Tiến độ 5
  // ==========================================

  searchReferenceValues: async (type, keyword = '') => {
    const response = await axiosClient.get(`/candidates/reference-values/search`, {
      params: { type, keyword }
    });
    return response.data;
  },

  // request: { type, name, requestType: 'CREATE'|'EDIT'|'DELETE', targetReferenceValueId? }
  suggestReferenceValue: async (userId, request) => {
    const response = await axiosClient.post(`/candidates/${userId}/reference-values/suggestions`, request);
    return response.data;
  },

  // ==========================================
  // 6. HỌC VẤN (EDUCATIONS) — [MỚI]
  // ==========================================

  getEducations: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/educations`);
    return response.data;
  },
  createEducation: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/educations`, payload);
    return response.data;
  },
  updateEducation: async (userId, educationId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/educations/${educationId}`, payload);
    return response.data;
  },
  deleteEducation: async (userId, educationId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/educations/${educationId}`);
    return response.data;
  },

  // ==========================================
  // 7. KINH NGHIỆM LÀM VIỆC (EXPERIENCES) — [MỚI]
  // ==========================================

  getExperiences: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/experiences`);
    return response.data;
  },
  createExperience: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/experiences`, payload);
    return response.data;
  },
  updateExperience: async (userId, experienceId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/experiences/${experienceId}`, payload);
    return response.data;
  },
  deleteExperience: async (userId, experienceId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/experiences/${experienceId}`);
    return response.data;
  },

  // ==========================================
  // 8. DỰ ÁN (PROJECTS) — [MỚI]
  // ==========================================

  getProjects: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/projects`);
    return response.data;
  },
  createProject: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/projects`, payload);
    return response.data;
  },
  updateProject: async (userId, projectId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/projects/${projectId}`, payload);
    return response.data;
  },
  deleteProject: async (userId, projectId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/projects/${projectId}`);
    return response.data;
  },

  // ==========================================
  // 9. CHỨNG CHỈ (CERTIFICATIONS) — [MỚI]
  // ==========================================

  getCertifications: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/certifications`);
    return response.data;
  },
  createCertification: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/certifications`, payload);
    return response.data;
  },
  updateCertification: async (userId, certificationId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/certifications/${certificationId}`, payload);
    return response.data;
  },
  deleteCertification: async (userId, certificationId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/certifications/${certificationId}`);
    return response.data;
  },

  // ==========================================
  // 10. HOẠT ĐỘNG (ACTIVITIES) — [MỚI]
  // ==========================================

  getActivities: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/activities`);
    return response.data;
  },
  createActivity: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/activities`, payload);
    return response.data;
  },
  updateActivity: async (userId, activityId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/activities/${activityId}`, payload);
    return response.data;
  },
  deleteActivity: async (userId, activityId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/activities/${activityId}`);
    return response.data;
  },

  // ==========================================
  // 11. GIẢI THƯỞNG (AWARDS) — [MỚI]
  // ==========================================

  getAwards: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/awards`);
    return response.data;
  },
  createAward: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/awards`, payload);
    return response.data;
  },
  updateAward: async (userId, awardId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/awards/${awardId}`, payload);
    return response.data;
  },
  deleteAward: async (userId, awardId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/awards/${awardId}`);
    return response.data;
  },

  // ==========================================
  // 12. SỞ THÍCH (HOBBIES) — [MỚI]
  // ==========================================

  getHobbies: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/hobbies`);
    return response.data;
  },
  createHobby: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/hobbies`, payload);
    return response.data;
  },
  updateHobby: async (userId, hobbyId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/hobbies/${hobbyId}`, payload);
    return response.data;
  },
  deleteHobby: async (userId, hobbyId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/hobbies/${hobbyId}`);
    return response.data;
  },

  // ==========================================
  // 13. NGÔN NGỮ (LANGUAGES) — [MỚI]
  // ==========================================

  getLanguages: async (userId) => {
    const response = await axiosClient.get(`/candidates/${userId}/languages`);
    return response.data;
  },
  createLanguage: async (userId, payload) => {
    const response = await axiosClient.post(`/candidates/${userId}/languages`, payload);
    return response.data;
  },
  updateLanguage: async (userId, languageId, payload) => {
    const response = await axiosClient.put(`/candidates/${userId}/languages/${languageId}`, payload);
    return response.data;
  },
  deleteLanguage: async (userId, languageId) => {
    const response = await axiosClient.delete(`/candidates/${userId}/languages/${languageId}`);
    return response.data;
  },

  // ==========================================
  // 14. QUẢN LÝ ĐƠN ỨNG TUYỂN (giữ nguyên như cũ)
  // ==========================================

  getMyApplications: async (userId, page = 0, size = 10) => {
    const response = await axiosClient.get(`/applications/candidates/${userId}?page=${page}&size=${size}`);
    return response.data;
  },

  extractCv: async (userId, file) => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await axiosClient.post(`/candidates/${userId}/cvs/extract`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },
};

export default candidateService;

File: backend-core/application/src/main/java/com/worklify/application/candidate/service/CandidateService.java
package com.worklify.application.candidate.service;

import com.worklify.application.candidate.dto.*;
import com.worklify.application.common.dto.PageResponse;
import com.worklify.application.referencedata.dto.ReferenceValueResponse;
import com.worklify.application.referencedata.dto.SuggestionRequest;
import com.worklify.application.referencedata.dto.SuggestionResponse;
import com.worklify.domain.common.DomainPageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface CandidateService {
    CandidateProfileResponse createProfile(Long userId, CandidateProfileRequest request);
    CandidateProfileResponse updateProfile(Long userId, CandidateProfileRequest request);
    CandidateProfileResponse getProfileByUserId(Long userId);

    CandidateProfileResponse uploadAvatar(Long userId, MultipartFile file);

    CvDocumentResponse uploadCv(Long userId, MultipartFile file);
    CvDocumentResponse saveGeneratedCv(Long userId, String rawText);
    CvDocumentResponse getLatestGeneratedCv(Long userId);
    List<CvDocumentResponse> getCvsByUserId(Long userId);
    void deleteCv(Long userId, Long cvId);

    List<CandidateSkillResponse> getSkillsByUserId(Long userId);
    CandidateSkillResponse createSkill(Long userId, CandidateSkillRequest request);
    CandidateSkillResponse updateSkill(Long userId, Long skillId, CandidateSkillRequest request);

    void addSkillToCandidate(Long userId, Long skillId);
    void removeSkillFromCandidate(Long userId, Long skillId);

    // [MỚI] Dropdown search + đề xuất bổ sung, dùng chung cho SKILL/LANGUAGE/...
    List<ReferenceValueResponse> searchReferenceValues(String type, String keyword);
    SuggestionResponse suggestReferenceValue(Long userId, SuggestionRequest request);

    CvDocumentResponse renameCv(Long userId, Long cvId, String newName);
    PageResponse<CandidateProfileResponse> searchCandidates(String keyword, DomainPageable pageable);
    CvDocumentResponse getCvDetail(Long userId, Long cvId);
    CvDocumentResponse updateGeneratedCv(Long userId, Long cvId, String rawText);

    CvDocumentResponse uploadCvThumbnail(Long userId, Long cvId, MultipartFile file);

    CvDocumentResponse uploadCvDigitalPdf(Long userId, Long cvId, MultipartFile file);

    List<EducationResponse> getEducationsByUserId(Long userId);
    EducationResponse createEducation(Long userId, EducationRequest request);
    EducationResponse updateEducation(Long userId, Long educationId, EducationRequest request);
    void deleteEducation(Long userId, Long educationId);

    // Experience
    List<ExperienceResponse> getExperiencesByUserId(Long userId);
    ExperienceResponse createExperience(Long userId, ExperienceRequest request);
    ExperienceResponse updateExperience(Long userId, Long experienceId, ExperienceRequest request);
    void deleteExperience(Long userId, Long experienceId);

    // Project
    List<ProjectResponse> getProjectsByUserId(Long userId);
    ProjectResponse createProject(Long userId, ProjectRequest request);
    ProjectResponse updateProject(Long userId, Long projectId, ProjectRequest request);
    void deleteProject(Long userId, Long projectId);

    // Certification
    List<CertificationResponse> getCertificationsByUserId(Long userId);
    CertificationResponse createCertification(Long userId, CertificationRequest request);
    CertificationResponse updateCertification(Long userId, Long certificationId, CertificationRequest request);
    void deleteCertification(Long userId, Long certificationId);

    // Activity
    List<ActivityResponse> getActivitiesByUserId(Long userId);
    ActivityResponse createActivity(Long userId, ActivityRequest request);
    ActivityResponse updateActivity(Long userId, Long activityId, ActivityRequest request);
    void deleteActivity(Long userId, Long activityId);

    // Award
    List<AwardResponse> getAwardsByUserId(Long userId);
    AwardResponse createAward(Long userId, AwardRequest request);
    AwardResponse updateAward(Long userId, Long awardId, AwardRequest request);
    void deleteAward(Long userId, Long awardId);

    // Hobby
    List<HobbyResponse> getHobbiesByUserId(Long userId);
    HobbyResponse createHobby(Long userId, HobbyRequest request);
    HobbyResponse updateHobby(Long userId, Long hobbyId, HobbyRequest request);
    void deleteHobby(Long userId, Long hobbyId);

    // Language
    List<LanguageResponse> getLanguagesByUserId(Long userId);
    LanguageResponse createLanguage(Long userId, LanguageRequest request);
    LanguageResponse updateLanguage(Long userId, Long languageId, LanguageRequest request);
    void deleteLanguage(Long userId, Long languageId);

    // Lấy layout hiện tại; tự khởi tạo mặc định nếu candidate mở ProfilePage lần đầu
    List<ProfileLayoutItemResponse> getProfileLayout(Long userId);

    // Kéo-thả đổi vị trí nhiều block cùng lúc
    List<ProfileLayoutItemResponse> reorderProfileLayout(Long userId, ProfileLayoutReorderRequest request);

    // Ẩn/hiện 1 block
    ProfileLayoutItemResponse toggleBlockVisibility(Long userId, String blockType, boolean visible);

    List<CandidateSkillResponse> reorderSkills(Long userId, SkillReorderRequest request);

    CandidateProfileFullResponse getFullProfile(Long userId);

    ParsedCvResponse extractCvFromFile(Long userId, MultipartFile file);
}